### PR TITLE
Reorganize types ns functions

### PR DESCRIPTION
**Pull request by sydney-runkle**
Refactor types namespace modification functions to highlight 2 different use cases.

I'm in no rush to merge this - would love some other ideas about how we could restructure this to make the most sense internally.

One idea - we could break out another function that adds `cls.__name__` to the namespace.

### CODE REVIEW

File: pydantic/_internal/_dataclasses.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ def complete_dataclass(
         )
 
     if types_namespace is None:
-        types_namespace = _typing_extra.get_cls_types_namespace(cls)
+        types_namespace = _typing_extra.merge_cls_and_parent_ns(cls)
```
Comment (sydney-runkle) [inline pydantic/_internal/_dataclasses.py:118]: So we could change this to ` _typing_extra.get_module_ns_of(cls).copy()`, but I think we should instead refactor how we handle namespaces for dataclasses in general to be more consistent with that of models.

See https://github.com/pydantic/pydantic/blob/e13b15edbe6cf90c603fbda1246b9156764c810f/pydantic/dataclasses.py#L318-L321

**Review by dmontagu — APPROVED** [review]
Generally looks good to me, though IIRC @alexmojaki had good ideas about removing the .copy (not sure how much that matters).

**Comment by sydney-runkle** [comment]
> Generally looks good to me, though IIRC @alexmojaki had good ideas about removing the .copy (not sure how much that matters).

Yes - definitely inclined to do this. Going to submit another PR to refactor in that regard!

File: pydantic/_internal/_typing_extra.py [hunk 2]
```diff
@@ -203,24 +203,25 @@ def parent_frame_namespace(*, parent_depth: int = 2, force: bool = False) -> dic
...
-    return ns
+            return {}
+    return {}
 
 
-def get_cls_types_namespace(cls: type[Any], parent_namespace: dict[str, Any] | None = None) -> dict[str, Any]:
-    ns = add_module_globals(cls, parent_namespace)
+def merge_cls_and_parent_ns(cls: type[Any], parent_namespace: dict[str, Any] | None = None) -> dict[str, Any]:
+    ns = get_module_ns_of(cls).copy()
+    ns.update(parent_namespace or {})
```
Comment (Viicos) [inline pydantic/_internal/_typing_extra.py:224]: I would do

```suggestion
    if parent_namespace is not None:
        ns.update(parent_namespace)
```

More stylistic than anything else, although this improves performance by a factor of 10 when `parent_namespace` is `None` (probably not relevant, but still good to have :) ).

**Review by Viicos — APPROVED** [review]

**Pull request merged**